Multi-chain Proposal

okxwallet opened the pull request:
Ref: [Multi-chain Proposal](https://dialectlabs.notion.site/EXTERNAL-Multichain-proposal-for-OKX-16ad2347008049d4a1e1bef19890946a)

zhelezkov commented on src/api/actions-spec.ts:
```diff
@@ -54,3 +54,8 @@ export interface ActionsSpecPostRequestBody
+
+export interface Options {
+  // For multi-chain, see https://github.com/ChainAgnostic/CAIPs
+  blockchainIds?: string;
```
I understand we didn't want to support multiple ids yet, but maybe we should use `string[]` type to avoid breaking change in future?

zhelezkov reviewed (changes requested):
Hey, thanks for PR!

In the requirements above you mention that
> Blink client decides whether to unfurl an action based on the X-Blockchain-Ids header

Could you please elaborate a bit maybe with code example, how this requirement satisfied with this PR? e.g. how wallet can decide not to unfurl an action?
Or you don't wanna control actions unfurling based on this header for now?

okxwallet commented:
> Reference in ne

The cited section refers to the proposal details. Would it be more appropriate for the decision regarding unfurl to be made by the wallet? Should it not be included in the blinks repository?

tsmbl commented:
> > Reference in ne
> 
> The cited section refers to the proposal details. Would it be more appropriate for the decision regarding unfurl to be made by the wallet? Should it not be included in the blinks repository?
> 
> Would this be acceptable?
> 
> ```ts
> export interface ActionAdapter {
>   connect: (context: ActionContext) => Promise<string | null>;
>   signTransaction: (
>     tx: string,
>     context: ActionContext,
>   ) => Promise<{ signature: string } | { error: string }>;
>   confirmTransaction: (
>     signature: string,
>     context: ActionContext,
>   ) => Promise<void>;
>   shouldRenderBlink?: (
>     options: Options
>   ) => Promise<boolean>;
> }
> 
> export interface Options {
>   // For multi-chain, see https://github.com/ChainAgnostic/CAIPs
>   blockchainIds?: string[];
> }
> ```

Hey @okxwallet, I think this looks right conceptually. The wallet should decide whether to render by implementing a function like `shouldRenderBlink`. I suggest some changes:

1. Rename `shouldRenderBlink` to be more abstract and use fewer nouns. For example, name or `isCompatible`.
2. Move `Options` as an attribute in `ActionContext` and provide it in the `isCompatible` callback. This way, compatibility information is available in a broader context and aligns with the structure of the interface.
3. For the context: we are discussing versioning of blink clients and action providers. I suspect, in the future the compatibility  decision may rely on at least two attributes: blockchains and versions, so I would try to be extensible from day 1. 

Here is my vision on how your proposal can be updated to reflect the thoughts above

```
export interface ActionContext {
  originalUrl: string;
  action: Action;
  actionType: 'trusted' | 'malicious' | 'unknown';
  triggeredLinkedAction: ActionComponent;
  actionCompatibility: ActionCompatibility;
}
export interface ActionCompatibility {
  blockchainIds?: string[];
}

export interface ActionAdapter {
  connect: (context: ActionContext) => Promise<string | null>;
  signTransaction: (
    tx: string,
    context: ActionContext,
  ) => Promise<{ signature: string } | { error: string }>;
  confirmTransaction: (
    signature: string,
    context: ActionContext,
  ) => Promise<void>;
  isCompatible?: (context: ActionContext) => Promise<boolean>;
}
```

wdyt? cc @zhelezkov

okxwallet commented:
> Hey @okxwallet, I think this looks right conceptually. The wallet should decide whether to render by implementing a function like `shouldRenderBlink`. I suggest some changes:
> 
> 1. Rename `shouldRenderBlink` to be more abstract and use fewer nouns. For example, name or `isCompatible`.
> 2. Move `Options` as an attribute in `ActionContext` and provide it in the `isCompatible` callback. This way, compatibility information is available in a broader context and aligns with the structure of the interface.
> 3. For the context: we are discussing versioning of blink clients and action providers. I suspect, in the future the compatibility decision may rely on at least two attributes: blockchains and versions, so I would make interfaces extensible from day 1.
> 
> Here is my vision on how your proposal can be updated to reflect the thoughts above
> 
> ```ts
> export interface ActionContext {
>   originalUrl: string;
>   action: Action;
>   actionType: 'trusted' | 'malicious' | 'unknown';
>   triggeredLinkedAction: ActionComponent;
>   actionCompatibility: ActionCompatibility;
> }
> export interface ActionCompatibility {
>   blockchainIds?: string[];
> }
> 
> export interface ActionAdapter {
>   connect: (context: ActionContext) => Promise<string | null>;
>   signTransaction: (
>     tx: string,
>     context: ActionContext,
>   ) => Promise<{ signature: string } | { error: string }>;
>   confirmTransaction: (
>     signature: string,
>     context: ActionContext,
>   ) => Promise<void>;
>   isCompatible?: (context: ActionContext) => Promise<boolean>;
> }
> ```
> 
> wdyt? cc @zhelezkov

Hey @tsmbl. Great thoughts
We totally agree that compatibility information should available in a broader context and aligns with the structure of the interface. However, when we implemented it, we found that some state in the `ActionContext` is hard to get.

Specifically, the `triggeredLinkedAction`(`ActionComponent`) .
The `ActionComponent` type is defined to bind with a specific component in the blink, e.g. a button or an input:
```typescript
export class ActionComponent {
  private parameterValue: Record<string, string> = {};

  constructor(
    private _parent: Action,
    private _label: string,
    private _href: string,
    private _parameters?: Parameter[]
  ) {}
  // ...
}
```
In other words, it happens and binds states **during the render**, but we think that both of us would agree that `isCompatible` should be called **before the rendering**.

We've updated our PR with other your suggestions:
- rename `shouldRenderBlink` to `isCompatible`
- add `ActionCompatibility` type definition

Please tell us your thoughts on this and we'd happy to update our proposal and code.

alex-tsymbal commented:
Hey @okxwallet, thanks for the update

Agreed that `triggeredLinkedAction` is not expected to present when`isCompatible` is called. 

As an idea, we could omit the `triggeredLinkedAction` specifically inside `isCompatible`, wdyt?

```
isCompatible?: (
  context: Omit<ActionContext, 'triggeredLinkedAction'>,
) => Promise<boolean>;
```

tsmbl commented:
Hey @okxwallet, thanks for the update

Agreed that `triggeredLinkedAction` is not expected to present when`isCompatible` is called. 

As an idea, we could omit the `triggeredLinkedAction` specifically inside `isCompatible`, wdyt?

```
isCompatible?: (
  context: Omit<ActionContext, 'triggeredLinkedAction'>,
) => Promise<boolean>;
```

okxwallet commented:
@tsmbl Updated

zhelezkov commented:
Hey all, I have some opinions on changes:
- I would propose to make better abstraction not related to compatibility, e.g. `ActionMetadata` and expose it as part of `Action` object. This way we can avoid unnecessary changes everywhere in the code
- `isCompatible` -> `isSupported`, prob better wording, more abstract one. I think it might be useful in future for wallets understand if they supported specific `Action` or not based on various checks

In conscious of time I've wrapped my proposed changes in a commit https://github.com/dialectlabs/blinks/pull/12/commits/dfccc733da7359206831f09e7e8b889309150e79
Please feel free to revert or add smth!

zhelezkov reviewed (approved):
with my latest changes I'm ok to merge it

okxwallet commented:
@zhelezkov Great, we're good with this, let's ship it!
